packages/@aws-cdk/aws-appsync/test/appsync-schema.test.ts

```diff
@@ -272,4 +283,108 @@ describe('testing schema definition mode `file`', () => {
       api.addSubscription('blah', new appsync.ResolvableField({ returnType: t.string }));
     }).toThrowError('Unable to add subscription. Schema definition mode must be CODE. Received: FILE');
   });
+});
+
+describe('testing schema definition mode `s3`', () => {
+
+  test('definition mode `s3` produces correct output', () => {
+    // GIVEN
+    const schemaAsset = new assets.Asset(stack, 'Schema', {
+      path: join(__dirname, 'appsync.test.graphql'),
+    });
+    // WHEN
+    new appsync.GraphqlApi(stack, 'API', {
+      name: 'demo',
+      schema: appsync.Schema.fromS3Location(schemaAsset.s3ObjectUrl),
+    });
+
+    // THEN
+    expect(stack).toHaveResourceLike('AWS::AppSync::GraphQLSchema', {
+      DefinitionS3Location: {
+        'Fn::Join': ['', ['s3://',
+          { Ref: 'AssetParametersd915eeba150016e6e51f327f3d1f8a3701dfb7456e8dc17426c8960cecb2d9beS3BucketCB5CC368' },
+          '/',
+          {
+            'Fn::Select': [0, {
+              'Fn::Split': ['||',
+                { Ref: 'AssetParametersd915eeba150016e6e51f327f3d1f8a3701dfb7456e8dc17426c8960cecb2d9beS3VersionKeyB59D7BBD' }],
+            }],
+          },
+          {
+            'Fn::Select': [1, {
+              'Fn::Split': ['||'],
+            }],
+          }]],
+      },
+    });
+  });
+
+  test('definition mode `s3` errors when addType for object is called', () => {
+    // WHEN
+    const api = new appsync.GraphqlApi(stack, 'API', {
+      name: 'demo',
+      schema: appsync.Schema.fromS3Location('garbage'),
+    });
+
+    // THEN
+    expect(() => {
+      api.addType(new appsync.ObjectType('blah', {
+        definition: { fail: t.id },
+      }));
+    }).toThrowError('API cannot add type because schema definition mode is not configured as CODE.');
+  });
+
+  test('definition mode `s3` errors when addType for interface is called', () => {
+    // WHEN
+    const api = new appsync.GraphqlApi(stack, 'API', {
+      name: 'demo',
+      schema: appsync.Schema.fromS3Location('garbage'),
+    });
+
+    // THEN
+    expect(() => {
+      api.addType(new appsync.InterfaceType('blah', {
+        definition: { fail: t.id },
+      }));
+    }).toThrowError('API cannot add type because schema definition mode is not configured as CODE.');
+  });
```

Comment:
These 2 tests are exactly the same, no?

Reply:
An interface type is technically different than an object type but the validation is the same.. ill delete the interface one and rename the test to test against `IntermediateType`

Reply:
Let me ask this 🙂.

Is there anything that you can pass to `. addType()` there that would _not_ throw an exception?

Reply:
ah yes that is very true